extractor/docker/docker.go

```diff
@@ -257,29 +312,127 @@ func (d Extractor) extractLayerFiles(layerCh chan layer, errCh chan error, ctx c
 	return nil
 }
 
-func (d Extractor) extractLayerWorker(dig digest.Digest, r *registry.Registry, ctx context.Context, image registry.Image, errCh chan error, layerCh chan layer) {
-	var rc io.Reader
-	// Use cache
-	rc = d.Cache.Get(string(dig))
-	if rc == nil {
-		// Download the layer.
-		layerRC, err := r.DownloadLayer(ctx, image.Path, dig)
+func (d Extractor) extractLayerWorker(dig digest.Digest, r *registry.Registry, ctx context.Context, image registry.Image, errCh chan error, layerCh chan layer, filenames []string) {
+	var tarContent bytes.Buffer
+	var cacheContent []byte
+	found, _ := d.Cache.Get(kvtypes.GetItemInput{
+		BucketName: LayerTarsBucket,
+		Key:        string(dig),
+		Value:      &cacheContent,
+	})
+
+	if found {
+		b, err := extractTarContent(cacheContent)
+		if err != nil || len(b) <= 0 {
+			found = false
+		}
+
+		_, err = tarContent.Write(b)
+		if err != nil {
+			found = false
+		}
+	}
+
+	var cacheBuf bytes.Buffer
+	if !found {
+		rc, err := r.DownloadLayer(ctx, image.Path, dig)
 		if err != nil {
 			errCh <- xerrors.Errorf("failed to download the layer(%s): %w", dig, err)
 			return
 		}
+		defer rc.Close()
 
-		rc, err = d.Cache.Set(string(dig), layerRC)
+		// read the incoming gzip from the layer
+		gzipReader, err := gzip.NewReader(rc)
 		if err != nil {
-			log.Print(err)
+			errCh <- xerrors.Errorf("could not init gzip reader: %w", err)
+			return
+		}
+		defer gzipReader.Close()
+
+		tarReader := tar.NewReader(io.TeeReader(gzipReader, &tarContent))
+
+		if len(filenames) > 0 {
+			if cacheBuf, err = getFilteredTarballBuffer(tarReader, filenames); err != nil {
+				errCh <- err
+				return
+			}
 		}
+
+		d.storeLayerInCache(cacheBuf, dig)
+	}
+
+	layerCh <- layer{ID: dig, Content: ioutil.NopCloser(&cacheBuf)}
```

Comment:
When the layer id hits the cache, will `Content` be empty? This is because we wrote a cache content to `tarContent`. It seems to introduce a bug. I'm sorry if I misunderstand something.
https://github.com/aquasecurity/fanal/pull/51/files#diff-1ae53d6fc729587121b603c3ace40f24R330

Reply:
Yes, you're correct. This indeed is a bug and was a result because of a weak test assertion: https://github.com/aquasecurity/fanal/pull/51/files/f22c0ed94fb198ad800a480303e563615458d7bc#diff-4d409e1b4160e28284809dc87cfddec4R537 I have fixed this here: https://github.com/aquasecurity/fanal/pull/51/commits/8e233b311ac7b1ced61a64346b31ba1ef952b039